src/requestLib.js

```diff
@@ -0,0 +1,134 @@
+var http = require('http');
+var constants = require('../config/constants');
+var Utils = require('./utils');
+var keepAliveAgent = new http.Agent({
+  keepAlive: true
+});
+var RdGlobalConfig = constants.RdGlobalConfig;
+
+var RequestLib = {
+  /**
+   * Method to perform the request on behalf of the client
+   * @param {{request: Object, furtherRequestOptions: Object}} params 
+   * @param {http.IncomingMessage} clientRequest 
+   * @param {Number} retries 
+   */
+  _makeRequest: function (params, clientRequest, retries) {
+    return new Promise(function (resolve, reject) {
+      var requestOptions = Object.assign({}, params.furtherRequestOptions, {
+        agent: keepAliveAgent
+      });
+
+      // Adding a custom header for usage and debugging purpose at BrowserStack
+      requestOptions.headers['X-Requests-Debugger'] = clientRequest.id;
+
+      // Initialize the request to be fired on behalf of the client
+      var request = http.request(requestOptions, function (response) {
+        var responseToSend = {
+          statusCode: response.statusCode,
+          headers: response.headers,
+          data: []
+        };
+
+        response.on('data', function (chunk) {
+          responseToSend.data.push(chunk);
+        });
+
+        response.on('end', function () {
+          responseToSend.data = Buffer.concat(responseToSend.data).toString();
+          resolve(responseToSend);
+        });
+      });
+
+      // Log the request that will be initiated on behalf of the client
+      request.on('finish', function () {
+        RdGlobalConfig.ReqLogger.info('Tool Request - Retries Left: ' + retries, clientRequest.method + ' ' + clientRequest.url,
+          false,
+          Object.assign({}, params.furtherRequestOptions, {
+            data: Buffer.concat(params.request.data).toString()
+          }),
+          clientRequest.id);
+      });
+
+      // Capture any error scenarios while making the request on behalf of the client
+      request.on('error', function (err) {
+        reject({
+          message: err,
+          customTopic: 'Tool Request - Retries Left: ' + retries
+        });
+      });
+
+      // Set a hard timeout for the request being initiated.
+      request.setTimeout(constants.CLIENT_REQ_TIMEOUT, function () {
```

Comment:
If the user has set different request timeout from its side and it is greater than `CLIENT_REQ_TIMEOUT` then how are we handling that case?

Reply:
Almost all the requests from client bindings get timedout before the value there. Should we set a high value for this? Or up the default value?

Reply:
May be bumping this value or asking for the request timeout to set for each request.

Reply:
How much shall I bump this to? 
Setting it for each request doesn't seem feasible. The user might have to modify his/her scripts to incorporate that.

Reply:
May be ask as an arg to proxy for the request timeout from the user.

Reply:
^ That can be done. But I am just wondering around as to the requirement of making it configurable. A request not returning any response after 2 or 3 minutes would have anyway died at the client binding ( in case the user is not trying something else). What do you think?

Reply:
Yes, true, but i think in our constants it is `50 seconds`.

Reply:
I have set the limit to 260 seconds. Also made it configurable via CLI args.

Reply:
Discussed f2f. Resolving.